python/lsst/ctrl/mpexec/mock_task.py

```diff
@@ -0,0 +1,217 @@
+        Data butler instance.
+    quantum : `~lsst.daf.butler.Quantum`
+        Execution quantum.
+
+    Notes
+    -----
+    This implementation overrides get method to try to retrieve dataset from a
+    mock dataset type if it exists. Get method always returns a dictionary.
+    Put method stores the data with a mock dataset type, but also registers
+    DatasetRef with registry using original dataset type.
+    """
+
+    def __init__(self, butler: Butler, quantum: Quantum):
+        super().__init__(butler, quantum)
+        self.butler = butler
+
+    @classmethod
+    def mockDatasetTypeName(cls, datasetTypeName: str) -> str:
+        """Make mock dataset type name from actual dataset type name."""
+        return "_mock_" + datasetTypeName
+
+    def _get(self, ref: DatasetRef) -> Any:
+        # docstring is inherited from the base class
+        if isinstance(ref, DeferredDatasetRef):
+            ref = ref.datasetRef
+        datasetType = ref.datasetType
+
+        typeName, component = datasetType.nameAndComponent()
+        if component is not None:
+            mockDatasetTypeName = self.mockDatasetTypeName(typeName)
+        else:
+            mockDatasetTypeName = self.mockDatasetTypeName(datasetType.name)
+
+        try:
+            mockDatasetType = self.butler.registry.getDatasetType(mockDatasetTypeName)
+            ref = DatasetRef(mockDatasetType, ref.dataId)
+            data = self.butler.get(ref)
+        except KeyError:
+            data = super()._get(ref)
+
+        if not isinstance(data, dict):
+            data = {
+                "ref": {
+                    "dataId": {key.name: ref.dataId[key] for key in ref.dataId.keys()},
+                    "datasetType": ref.datasetType.name,
+                },
+                "type": get_full_type_name(type(data)),
+            }
+        if component is not None:
+            data.update(component=component)
+        return data
+
+    def _put(self, value: Any, ref: DatasetRef):
+        # docstring is inherited from the base class
+
+        mockDatasetType = self.registry.getDatasetType(self.mockDatasetTypeName(ref.datasetType.name))
+        mockRef = DatasetRef(mockDatasetType, ref.dataId)
+        value.setdefault("ref", {}).update(datasetType=mockDatasetType.name)
+        self.butler.put(value, mockRef)
+
+        # also "store" non-mock refs
+        self.registry._importDatasets([ref])
+
+    def _checkMembership(self, ref: Union[List[DatasetRef], DatasetRef], inout: set):
+        # docstring is inherited from the base class
+        return
+
+
+class MockPipelineTaskConfig(PipelineTaskConfig, pipelineConnections=PipelineTaskConnections):
+
+    failCondition = Field(
+        dtype=str,
+        default="",
+        doc=(
+            "Condition on DataId to raise an exception. String expression which includes attributes of "
+            "quantum DataId using a syntax of daf_butler user expressions (e.g. 'visit = 123')."
```

Comment:
I'm not sure I understood this comment. The expression is for the _quantum_ DataId keys/values, not dataset DataIds. The expression can actually match more than one quantum if needed, it's OK to skip one or more keys from the expression.

Reply:
For example, I _think_ if you run a pipeline with an `instrument: HSC` key, you can say `-d visit=1228`.  Here I think you have to write `-d instrument='HSC' AND visit=1228`.

But now that I've said, that you don't have the check that makes it complain either - you'll just blindly use the visit number on any instrument you see, which is probably fine here.  So nevermind!

Reply:
Right, if you skip `instrument` here then it will trigger on any instrument, if this is not what you want then you have to give specific instrument too.